rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java

```diff
@@ -0,0 +1,897 @@
+        public Response actionWithLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI contextId,
+                                      @DefaultValue("false") @QueryParam("cancel") Boolean cancel) {
+            status = LRAStatus.Active;
+
+            server.stop(); //simulate a server crash
+
+            return getResult(cancel, contextId);
+        }
+        @LRA(value = LRA.Type.NESTED, end = false)
+        @PUT
+        @Path("nested")
+        public Response nestedLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI contextId,
+                                  @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentLRA,
+                                  @DefaultValue("false") @QueryParam("cancel") Boolean cancel) {
+            return getResult(cancel, contextId);
+        }
+        @LRA(value = LRA.Type.NESTED)
+        @PUT
+        @Path("nested-with-close")
+        public Response nestedLRAWithClose(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI contextId,
+                                           @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentId,
+                                           @DefaultValue("false") @QueryParam("cancel") Boolean cancel) {
+            return getResult(cancel, contextId);
+        }
+        @PUT
+        @Path("multiLevelNestedActivity")
+        @LRA(value = LRA.Type.MANDATORY, end = false)
+        public Response multiLevelNestedActivity(
+                @HeaderParam(LRA_HTTP_RECOVERY_HEADER) URI recoveryId,
+                @HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI nestedLRAId,
+                @QueryParam("nestedCnt") @DefaultValue("1") Integer nestedCnt) {
+            // invoke resources that enlist nested LRAs
+            String[] lras = new String[nestedCnt + 1];
+            lras[0] = nestedLRAId.toASCIIString();
+            IntStream.range(1, lras.length).forEach(i -> lras[i] = restPutInvocation(nestedLRAId,"nestedActivity", ""));
+
+            return Response.ok(String.join(",", lras)).build();
+        }
+        @PUT
+        @Path("nestedActivity")
+        @LRA(value = LRA.Type.NESTED, end = true)
+        public Response nestedActivity(@HeaderParam(LRA_HTTP_RECOVERY_HEADER) URI recoveryId,
+                                       @HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI nestedLRAId) {
+            return Response.ok(nestedLRAId.toASCIIString()).build();
+        }
+        @GET
+        @Path("status")
+        public Response getStatus() {
+            return Response.ok(status.name()).build();
+        }
+        @PUT
+        @Path("/complete")
+        @Complete
+        public Response complete(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI contextLRA,
+                                 @HeaderParam(LRA_HTTP_PARENT_CONTEXT_HEADER) URI parentLRA) {
+            if (acceptCount.getAndDecrement() <= 0) {
+                completeCount.incrementAndGet();
+                acceptCount.set(0);
```

Comment:
Yes, complete can be called more than once in a distributed system. Besides, checking for <= 0 is not affecting what the test is validating. If you study each test then you will find that a negative value has no effect on whether the test passes or fails. I can rework the tests and resources to guarantee that it will always be non-negative, but is that really worth the effort and is there much to be gained.

Reply:
ok, thanks